crates/bevy_diagnostic/src/diagnostic.rs

```diff
@@ -37,22 +37,30 @@ pub struct Diagnostic {
     history: VecDeque<DiagnosticMeasurement>,
     sum: f64,
     max_history_length: usize,
+    pub is_enabled: bool,
 }
 
 impl Diagnostic {
+    /// Add a new value as a [`DiagnosticMeasurement`]. Its timestamp will be [`Instant::now`].
     pub fn add_measurement(&mut self, value: f64) {
         let time = Instant::now();
-        if self.history.len() == self.max_history_length {
-            if let Some(removed_diagnostic) = self.history.pop_front() {
-                self.sum -= removed_diagnostic.value;
+        if self.max_history_length > 1 {
+            if self.history.len() == self.max_history_length {
+                if let Some(removed_diagnostic) = self.history.pop_front() {
+                    self.sum -= removed_diagnostic.value;
```

Comment:
A potential solution would be to switch form storing the sum to storing the simple moving average directly. An incremental SMA is calculated as

$$ SMA_t = \frac{ X_t + SMA_{t-1} \times (n-1) }{ n } $$

where the impact of any rounding error trends towards zero.

The main advantage of the incremental SMA over the typical and current $\frac{ X_t + X_{t-1} + \cdots + X_{t-n} }{ n }$ is not needing to store the historical readings, though, and it's perhaps useful to have the history, especially for discrete diagnostics rather than continuous.

Perhaps the best solution really is to separate discrete from continuous diagnostics.

Reply:
Strike that formula that's what I get for going off of memory. That's not a correct simple moving average; the SMA does require keeping track of the history. The correct incremental SMA formula is $$SMA_{k,\text{next}} = SMA_{k,\text{prev}} + \frac{1}{k}\left( p_{n+1} - p_{n-k+1} \right)$$. What I wrote above is some odd form of a weighted moving average where each sample has weight IIUC $1/(n+\text{age})$.